src/pydicom/pixels/decoders/base.py

```diff
@@ -0,0 +1,1980 @@
+        """Create a new runner for decoding data encoded as `tsyntax`.
+
+        Parameters
+        ----------
+        tsyntax : pydicom.uid.UID
+            The transfer syntax UID corresponding to the pixel data to be
+            decoded.
+        """
+        self._src: Buffer
+        self._src_type: str
+        self._opts: DecodeOptions = {
+            "transfer_syntax_uid": tsyntax,
+            "as_rgb": True,
+        }
+        self._decoders: dict[str, DecodeFunction] = {}
+        self._previous: tuple[str, DecodeFunction]
+
+        if self.transfer_syntax.is_encapsulated:
+            self.set_option("pixel_keyword", "PixelData")
+        else:
+            self.set_option("view_only", False)
+
+    @property
+    def bits_allocated(self) -> int:
+        """Return the expected number of bits allocated used by the data."""
+        if (value := self._opts.get("bits_allocated", None)) is not None:
+            return value
+
+        raise AttributeError("No value for 'bits_allocated' has been set")
+
+    @property
+    def bits_stored(self) -> int:
+        """Return the expected number of bits stored used by the data."""
+        if (value := self._opts.get("bits_stored", None)) is not None:
+            return value
+
+        raise AttributeError("No value for 'bits_stored' has been set")
+
+    @property
+    def columns(self) -> int:
+        """Return the expected number of columns in the data."""
+        if (value := self._opts.get("columns", None)) is not None:
+            return value
+
+        raise AttributeError("No value for 'columns' has been set")
+
+    def decode(self, index: int) -> bytes | bytearray:
+        """Decode the frame of pixel data at `index`.
+
+        Parameters
+        ----------
+        index : int
+            The index of the frame to be decoded.
+
+        Returns
+        -------
+        bytes | bytearray
+            The decoded frame of pixel data.
+        """
+        # For encapsulated data `self.src` should not be memoryview to avoid
+        #   creating a duplicate object in memory by the encapsulation functions
+        src = get_frame(
+            self.src,
+            index,
+            number_of_frames=self.number_of_frames,
+            extended_offsets=self.extended_offsets,
+        )
+
+        return self._decode_frame(src)
+
+    def _decode_frame(self, src: bytes) -> bytes | bytearray:
+        """Return a decoded frame of pixel data.
+
+        Parameters
+        ----------
+        src : bytes
+            An encoded frame of pixel data to be passed to the decoding plugins.
+
+        Returns
+        -------
+        bytes | bytearray
+            The decoded frame.
+        """
+        # If self._previous is not set then this is the first frame being decoded
+        # If self._previous is set and then the previously successful decoder
+        #   has failed while decoding a frame and we are trying the other decoders
```

Comment:
I don't understand the language of the second point here (the 'is set' case); I feel like the sentence was either supposed to continue, or the "is set and then" should be "is set, then...".

Reply:
It should be "is set, then the previously..."

Reply:
Fixed